### PR TITLE
Serve SPSP proxy and receipt verification requests on same port

### LINKED ISSUE #22: Run on single port
Running the SPSP proxy and receipts API on the same http server with a single port would simplify setup.
And with the `/balances/<id>:spend` API being removed in https://github.com/coilhq/receipt-verifier/pull/19, there's less concern about restricting access to any receipt verifier endpoints.

### PR DESCRIPTION
**Pull request by wilsonianb**
Rename `/receipts` endpoint to ~`/verifyReceipt`~ `/verify`

Resolves #22

### CODE REVIEW

File: src/routes/spsp.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          } else {
+            res.statusCode = 409
+            res.end()
+          }
+        } catch (err) {
+          res.statusCode = 409
+          res.end()
+        }
+      })
+    }.bind(this))
```
Comment (sharafian) [inline src/routes/spsp.ts:64]: bind isn't required for an arrow function

**Review by sharafian — APPROVED** [review]

**Review by sabinebertram — APPROVED** [review]
Any particular reason why the path is called `verifyReceipt` instead of `receipt` (because it was formerly `receipts`) or just `verify`?

**Comment by wilsonianb** [comment]
> Any particular reason why the path is called `verifyReceipt` instead of `receipt` (because it was formerly `receipts`) or just `verify`?

`verifyReceipt` seemed the clearest to me, but I like `verify`'s conciseness. I guess someone running this at an un-`receipt verifier`-y domain could map `/verifyReceipt` to `/verify` if they wanted

**Comment by wilsonianb** [comment]
Renamed as `/verify` :point_up_2:

File: package.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
   "scripts": {
     "build": "tsc",
     "start": "node dist/index.js",
-    "test": "jest"
+    "test": "jest --runInBand"
```
Comment (sublimator) [inline package.json:16]: 4 speed ?

Comment (wilsonianb) [reply package.json:16]: The route tests were racing to start a server on the same port, and I didn't bother figuring out how run everything concurrently except those two with each other.
I just realized I _could_ have route tests start the server on different ports, but I think I was already also getting spurious failures running the tests concurrently on a non-`mock` redis.

**Review by sharafian — APPROVED** [review]

File: src/routes/spsp.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        console.error(await endpointsRes.text())
+        ctx.throw(404)
+      }
+      spspEndpoint = await endpointsRes.text()
+    } else {
+      spspEndpoint = decodeURIComponent(ctx.req.url.substring(1))
+    }
+    const nonce = randomBytes(16)
+    const secret = generateReceiptSecret(ctx.config.receiptSeed, nonce)
+    ctx.proxyServer.once('proxyRes', (proxyRes: IncomingMessage, req: IncomingMessage, res: ServerResponse) => {
```
Comment (sentientwaffle) [inline src/routes/spsp.ts:40]: Won't this event handler (sometimes) mismatch with the `ctx` when multiple requests are in-flight? I'm not 100% sure, but it looks like the `proxyServer` object is shared between all middleware chains.

Comment (wilsonianb) [reply src/routes/spsp.ts]: :heavy_check_mark: you're right. Thanks for catching that. (I think the issue predates this pr.)
https://github.com/coilhq/receipt-verifier/pull/23/commits/77d1257997783c0a80125c2fc2c5aae209c2f1e3 swaps out `http-proxy` for an actual koa middleware.

File: src/routes/spsp.ts [hunk 4]
```diff
@@ -19,58 +20,51 @@ function resolvePointer (pointer: string): string {
...
-        'Receipt-Nonce': nonce.toString('base64'),
-        'Receipt-Secret': secret.toString('base64')
-      },
-      ignorePath: true,
-      selfHandleResponse: true,
-      target: resolvePointer(spspEndpoint)
-    })
+    ctx.state.nonce = randomBytes(16)
+    await next()
   }
```
[the remark this answers is not in the document]
Comment (wilsonianb) [reply src/routes/spsp.ts:38]: As is, koa returns `404`
https://github.com/coilhq/receipt-verifier/blob/77d1257997783c0a80125c2fc2c5aae209c2f1e3/src/routes/spsp.test.ts#L108-L116

File: src/routes/spsp.ts [hunk 5]
```diff
@@ -19,58 +20,51 @@ function resolvePointer (pointer: string): string {
   return url.href
 }
 
-export const router = new Router()
-
-router.get('/(.*)', async (ctx: Koa.Context) => {
+const spspProxySetup = () => async (ctx: Koa.Context, next: Koa.Next) => {
```
[the remark this answers is not in the document]
Comment (wilsonianb) [reply src/routes/spsp.ts:23]: It's not necessary. It semi-follows [this recommendation](https://github.com/koajs/koa/blob/master/docs/guide.md#middleware-options) without actually having options (while also not being a public middleware...)

File: src/routes/spsp.ts [hunk 6]
```diff
@@ -19,58 +20,51 @@ function resolvePointer (pointer: string): string {
...
+      proxyReqPathResolver: (ctx: Koa.Context) => {
+        return new URL(resolvePointer(ctx.state.spspEndpoint)).pathname
+      },
+      userResDecorator: async (proxyRes: IncomingMessage, proxyResData: Buffer, ctx: Koa.Context) => {
+        const data = JSON.parse(proxyResData.toString())
+        if (!data.receipts_enabled) {
+          ctx.throw(409)
+        }
+        await ctx.redis.cacheReceiptNonce(ctx.state.nonce.toString('base64'), ctx.state.spspEndpoint)
+        return proxyResData as Buffer
```
[the remark this answers is not in the document]
Comment (wilsonianb) [reply src/routes/spsp.ts]: It was doing nothing. I think at some point in development I had the function in such a way that it was complaining without the assertion.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**